[python][wheels] Use crossenv to generate cross compiled wheels

smaarn opened the pull request:
_Motivation:_  Generated wheels ship libraries were generated using the native `EXT_SUFFIX` value for native libraries. 
_Linked issues:_  Fixes https://github.com/SynoCommunity/spksrc/issues/3944

### Checklist
- [x] Build rule `all-supported` completed successfully (validated by building Python 3.7.7 and checking generated wheel contents)
- [x] Package upgrade completed successfully (validated by using Python 3.7.7 spk compiled with this)
- [x] New installation of package completed successfully

smaarn commented on mk/spksrc.python-wheel.mk:
```diff
@@ -25,7 +25,10 @@ PYTHONPATH = $(PYTHON_LIB_NATIVE):$(INSTALL_DIR)$(INSTALL_PREFIX)/$(PYTHON_LIB_D
 ### Python wheel rules
 build_python_wheel:
-	@$(RUN) PYTHONPATH=$(PYTHONPATH) $(HOSTPYTHON) -c "import setuptools;__file__='setup.py';exec(compile(open(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" $(BUILD_ARGS) bdist_wheel -b $(WORK_DIR)/wheelbuild -d $(WORK_DIR)/wheelhouse
+	@$(RUN) PYTHONPATH=$(PYTHONPATH) $(HOSTPYTHON) -c "import setuptools;\
```
I could have kept the one-liner but I thought it was becoming hard to read.

smaarn replied:
There is another (open) option which would be to patch the native python so that an environment variable (`TARGET_EXT_SUFFIX` ?) could override the `EXT_SUFFIX` variable computed upon compiling.

ymartin59 replied:
@smaarn I find it strange there is no standard "out-of-the-box" cross-compilation support for wheels with Python 3.

smaarn replied:
@ymartin59 I found this `crossenv`(https://pypi.org/project/crossenv/) python module which seems to be targetting cross compilation, the only thing is that it could mean some more changes to be made. From what I gathered it meant creating a shell script which would appropriately alias `python` to target a fake python which would be emulating the environment information for the target architecture.

ymartin59 replied:
@smaarn Correct. I have already been told about it. I will give another try to cross-compile and run borgbackup on current Python 3.6.10

smaarn commented on cross/python3/Makefile:
```diff
@@ -98,3 +98,4 @@ $(WORK_DIR)/python-cc.mk:
 	@echo PYTHON_INC_DIR=$(PYTHON_INC_DIR) >> $@
 	@echo PIP=$(PIP_NATIVE) >> $@
 	@echo CROSS_COMPILE_WHEELS=1 >> $@
+	@echo TARGET_EXT_SUFFIX=$(shell $(PYTHON_NATIVE) -c "from distutils.sysconfig import parse_makefile; print(parse_makefile('$(WORK_DIR)/Python-$(PKG_VERS)/Makefile')['EXT_SUFFIX']);") >> $@
```
We compute here the value which is expected to be used for the `EXT_SUFFIX` variable.

smaarn commented on cross/python3/Makefile:
```diff
@@ -98,3 +100,4 @@ $(WORK_DIR)/python-cc.mk:
 	@echo PYTHON_INC_DIR=$(PYTHON_INC_DIR) >> $@
 	@echo PIP=$(PIP_NATIVE) >> $@
 	@echo CROSS_COMPILE_WHEELS=1 >> $@
+	@echo TARGET_EXT_SUFFIX=$(shell $(PYTHON_NATIVE) -c "from distutils.sysconfig import parse_makefile; print(parse_makefile('$(WORK_DIR)/Python-$(PKG_VERS)/Makefile')['EXT_SUFFIX']);") >> $@
```
I believe this could be removed now.

hgy59 commented:
@smaarn you changed mk/spksrc.python-wheel.mk. Does this still work for python (2) modules?

smaarn commented:
> @smaarn you changed mk/spksrc.python-wheel.mk. Does this still work for python (2) modules?

@hgy59 I must confess I didn't worry about that part.

Two possibilities (if I continue down that path): 

1. I create a special mk for python3
2. I check python version (since I have that PYTHON_NATIVE environment variable available) I'm guessing it should be feasible

smaarn commented:
> > @smaarn you changed mk/spksrc.python-wheel.mk. Does this still work for python (2) modules?
> 
> @hgy59 I must confess I didn't worry about that part. All apologize.
> 
> Two possibilities (if I continue down that path):
> 
> 1. I create a special mk for python3
> 2. I check python version (since I have that PYTHON_NATIVE environment variable available) I'm guessing it should be feasible

@hgy59 I now updated this PR so that only wheels (not modules) relying on python 3 usage would be impacted by the usage of crossenv. Given the need to support python 2 I also reverted the removal of the renaming of the wheels.

smaarn commented:
I regenerated python 3 SPK based upon this here: https://github.com/smaarn/spksrc/releases/tag/python-3.6.11-2

I also generated the borgbackup ones. To be honest they are generated without any errors and install fine... But it seems to me that the spk is weird (meaning I don't see any borg binary or such a thing as mentioned in borg's documentation) so I'm a little bit at a loss on how to validate it. Any hint on how to validate the package ?

smaarn commented on cross/python3/Makefile:
```diff
@@ -82,12 +82,14 @@ python3_custom_install:
 python3_post_install: $(WORK_DIR)/python-cc.mk
 	mkdir -p $(PYTHON_LIB_CROSS)
 	cp -R $(HOSTPYTHON_LIB_NATIVE) $(PYTHON_LIB_CROSS)/../
+	@$(RUN) $(PYTHON_NATIVE) -m crossenv $(STAGING_INSTALL_PREFIX)/bin/python$(PKG_VERS_MAJOR).$(PKG_VERS_MINOR) $(WORK_DIR)/crossenv/
```
So here we basically run `crossenv` to setup the `activate` script at `$(WORKDIR)/crossenv/bin/activate` using the native python script and targetting the cross compiled python installation.

smaarn commented on cross/python3/Makefile:
```diff
@@ -82,12 +82,14 @@ python3_custom_install:
 
 $(WORK_DIR)/python-cc.mk:
 	@echo HOSTPYTHON=$(HOSTPYTHON) > $@
+	@echo CROSSENV=$(WORK_DIR)/crossenv/bin/activate >> $@
```
I configure the `CROSSENV` variable to target the generated `activate` script.

smaarn commented on mk/spksrc.python-wheel.mk:
```diff
@@ -25,7 +25,13 @@ PYTHONPATH = $(PYTHON_LIB_NATIVE):$(INSTALL_DIR)$(INSTALL_PREFIX)/$(PYTHON_LIB_D
 
 ### Python wheel rules
 build_python_wheel:
+ifeq ($(strip $(CROSSENV)),)
```
If `CROSSENV` environment variable isn't defined we'll assume we're going down the Python 2 road.
Otherwise we "source" the `$(CROSSENV)` script before running the build command using the `python` (aliased by the `$(CROSSENV)` script

smaarn commented on native/python3/Makefile:
```diff
@@ -29,10 +29,11 @@ PIP_NATIVE = $(WORK_DIR)/../../../native/$(PKG_NAME)/work-native/install/usr/loc
+	@$(MSG) Installing setuptools, pip, cffi and cross env
 	@$(RUN) wget https://bootstrap.pypa.io/ez_setup.py -O - | $(PYTHON)
 	@$(RUN) wget https://bootstrap.pypa.io/get-pip.py -O - | $(PYTHON)
 	@$(PIP) install "cffi==1.14.0"
```
We ensure that `crossenv` module is available for cross compilation of wheels

ymartin59 commented:
@smaarn Do we agree that this PR content is already included in #3953? If so, please close this one.

smaarn commented:
@ymartin59 indeed it is already included in #3953 .

The whole point of this PR was solely to be able to stabilize the existing 3.6 option.
